src/platform/Darwin/dnssd/DnssdHostNameRegistrar.mm

```diff
@@ -97,7 +84,18 @@
         }
         ReturnErrorOnFailure(error);
 
-        __auto_type onTimeout = ^{ wrapper(kDNSServiceErr_Timeout); };
+        __weak OnRegisterRecordCallback weakCallback = mOnRegisterRecordCallback;
+        __auto_type onTimeout = ^{
+            OnRegisterRecordCallback strongCallback = weakCallback;
+            if (!strongCallback) {
+                // Either we already fired or the object cleaned it up
+                return;
+            }
+
+            this->mOnRegisterRecordCallback = nullptr;
```

Comment:
Unfortunately, I don't think anything guarantees that `this` is valid just because `strongCallback` is not nil.

Concretely, if whoever called `register` still has a reference to the callback, that will keep it alive, but `this` could still get destroyed, no?

If we want to detect whether `this` is alive, what we have done elsewhere is the livenessTracker bits in NetworkMonitor, for example.